src/sitemap.jl

```diff
@@ -0,0 +1,101 @@
+# Note: Franklin.jl also implements sitemap generation:
+#
+#   https://github.com/tlienart/Franklin.jl/blob/f1f7d044dc95ba0d9f368a3d1afc233eb58a59cf/src/manager/sitemap_generator.jl#L51
+#
+# At some point it might be worth factoring the code into a small shared package.
+# Franklin's implementation is more general that this here, but it looks like it relies
+# of Franklin-specific globals, so it's not a trivial copy-paste into a separate package.
+
+const SITEMAP_URL_LIMIT = 50_000
+const SITEMAP_SIZE_LIMIT = 52_428_800
+
+function make_sitemap(;
+    sitemap_filename::AbstractString,
+    sitemap_root::AbstractString,
+    docs_root_directory::AbstractString,
+)
+    # Determine the list of sitemap URLs by finding all canonical URLs
+    sitemap_urls = find_sitemap_urls(; docs_root_directory, sitemap_root)
+    if length(sitemap_urls) == 0
+        @error "No sitemap URLs found"
+        return
+    end
+
+    sitemap_buffer = IOBuffer()
+    write(
+        sitemap_buffer,
+        """
+<?xml version="1.0" encoding="UTF-8"?>
+<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
+""",
+    )
+    for loc in sort(sitemap_urls)
+        write(sitemap_buffer, "<url><loc>$(loc)</loc></url>\n")
+    end
+    write(sitemap_buffer, "</urlset>\n")
+
+    # Sitemaps are limited to 50 000 URLs: https://www.sitemaps.org/protocol.html#index
+    # TODO: we could automatically split the sitemap up if it's bigger than that and
+    # generate a sitemap index.
+    if length(sitemap_urls) > SITEMAP_URL_LIMIT
+        @warn "Sitemap file has too many items: $(length(canonical_urls)) (limit: $(SITEMAP_URL_LIMIT))"
+    end
+    sitemap_bytes = take!(sitemap_buffer)
+    if length(sitemap_bytes) > SITEMAP_SIZE_LIMIT
+        @warn "Sitemap is too large: $(length(sitemap_bytes)) bytes (limit: $(SITEMAP_SIZE_LIMIT))"
+    end
```

Comment:
These should be errors instead, no? If the user specifically requests sitemap-generation, then we should hard-fail if we cannot generate something valid.

Reply:
I am happy either way. My only concern about turning those into errors is that you can have your builds fail arbitrarily and intermittently, since MultiDocumenter does not have control of the package upstreams. On the other hand, the sitemap being "invalid" is somewhat subjective -- it doesn't conform to the spec anymore, but we can still generate it (and I guess likely will be ignored by search engines), but we _can_ generate it.

At the same time, if we keep them as warnings, then no-one will notice them, and on those grounds it better to error. I think that in the future we want to automate splitting the sitemap up into multiple files if it gets too big -- that way we should be able to handle this case automatically.

Reply:
Updated this with the errors + soft limits.